### ISSUE TITLE
home assistant: Authentication error (you might have entered an invalid PIN code).  

### ISSUE THREAD

**Issue by loopway** [postlund/pyatv#539]
**Describe the bug**
add new integration appletv in home assistant 0.107.7 always results in error message "Authentication error (you might have entered an invalid PIN code)."

**To Reproduce**
every-time

**Expected behavior**
accept passcode for AirPlay protocol

**System Setup (please complete the following information):**
 - OS: centos 3.10.0-957.27.2.el7.x86_64
 - Python: 3.7.7
 - Home Assistant: 0.107.7
 - Docker: 19.03.8

**Additional context**
Tried multiple times. Even allowed AirPlay on AppleTV for "Everyone" in the settings.

DEBUG logs:

```
2020-03-29 01:14:28 DEBUG (MainThread) [pyatv.support.net] POST URL: http://10.0.0.31:7000/pair-setup-pin
2020-03-29 01:14:28 DEBUG (MainThread) [pyatv.support.net] -> Data[85]: b'62706c6973743030d201020304566d6574686f6454757365725370696e5f101036453830434242423141444634453341080d14191d0000000000000101000000000000000500000000000000000000000000000030'
2020-03-29 01:14:28 DEBUG (MainThread) [pyatv.support.net] <- Data[342]: b'62706c6973743030d20102030452706b5473616c744f1101008a781c4396a3b37c0da0df585b0aedd9de884ed080ff71d466a38f519247669a624586705ca2915fb68f73e2b0355f4c4794f9b055f2dfb85dec7fd09a7276bd83e3b102a6fc29c58e1b9c6d8302077a12f527d9ba376ed812f5d00c7887724b7983ee0d13d26a'...
2020-03-29 01:14:28 DEBUG (MainThread) [pyatv.airplay.srp] Client session key: b'a3f2cc146cb204145fcc746fa0d58dda513c6ddc9a90a4b793556729541e07158599661b788c9bf7'
2020-03-29 01:14:28 DEBUG (MainThread) [pyatv.airplay.srp] Client public: 67d0e6728b1208ad7db948b9d911a56d49016e686d7abe50d7255c1b1ba40718e2f3a70d0c5d0aa4fd9c7e91589eba4464858e2005aaf067477430e35f76afb699e44ef17df5000418e545f2798a3e8ee739c62acd871d328eb2ec08bf06b01e0bd527ea0c3f8ba909f6b04a0d5ae36addc2bb31dee2e4546fa66fc91d832d94c834b9f5d51a34534c63c6023a66f694fce9d662236a20aa32cff6a1f7de1a63eec69b7d8d62ca057f0f72ac9ac5f43dd17b02008cd319e27fefdd9b4ae39938236a21f3d99def8a688a0af507847d2d11d6a4097178064f4480a5aa88cf6fb149089a711ac273f92738a2fb52c9f0faac476e472ae863e4976856687a29b55c, proof: b'f5f0c4f3da6e42436eb296bec49c214796e470ba'
2020-03-29 01:14:28 DEBUG (MainThread) [pyatv.support.net] POST URL: http://10.0.0.31:7000/pair-setup-pin
2020-03-29 01:14:28 DEBUG (MainThread) [pyatv.support.net] -> Data[347]: b'62706c6973743030d20102030452706b5570726f6f664f11010067d0e6728b1208ad7db948b9d911a56d49016e686d7abe50d7255c1b1ba40718e2f3a70d0c5d0aa4fd9c7e91589eba4464858e2005aaf067477430e35f76afb699e44ef17df5000418e545f2798a3e8ee739c62acd871d328eb2ec08bf06b01e0bd527ea0c3f'...
2020-03-29 01:14:28 ERROR (MainThread) [custom_components.apple_tv.config_flow] Authentication problem
Traceback (most recent call last):
   File "/usr/local/lib/python3.7/site-packages/pyatv/support/__init__.py", line 29, in error_handler
     return await func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/pyatv/airplay/auth.py", line 49, in finish_authentication
     "step2", pk=binascii.unhexlify(pub_key), proof=binascii.unhexlify(key_proof)
   File "/usr/local/lib/python3.7/site-packages/pyatv/airplay/auth.py", line 69, in _send_plist
     raise AuthenticationError("{0} failed with code {1}".format(step, code))
 pyatv.exceptions.AuthenticationError: step2 failed with code 500
 The above exception was the direct cause of the following exception:
 Traceback (most recent call last):
   File "/config/custom_components/apple_tv/config_flow.py", line 290, in async_step_pair_with_pin
     await self.pairing.finish()
  File "/usr/local/lib/python3.7/site-packages/pyatv/airplay/pairing.py", line 70, in finish
     self.pin_code,
  File "/usr/local/lib/python3.7/site-packages/pyatv/support/__init__.py", line 37, in error_handler
     raise fallback(str(ex)) from ex
 pyatv.exceptions.PairingError: step2 failed with code 500
```

**Comment by postlund** [comment]
Interesting, which version of tvOS are you running?

**Comment by loopway** [comment]
13.3.1 and just updated it to 13.4 -> same errors.

**Comment by postlund** [comment]
Ok, good in some sense. I'm on 13.3.1 and added my device last friday, which worked fine. So it at least has nothing to do with 13.4. Did you by any chance try rebooting your Apple TV?

**Comment by ff4500** [comment]
Same issue here. Happens on both of my ATVs. Running 13.4 on an ATV 4k, and 13.3.1 (just updated to 13.4 to check if there was any difference) on an ATV 1080p. ATV scan sees both on the network just fine. PIN pops up on both just fine, but always throws:

```
Aborted

Authentication error. You might have entered an invalid PIN code or not have permission to pair via this protocol. Please check access settings on your Apple TV.
```

in Home Assistant. Integration is installed via HACS, if it matters.

**Comment by postlund** [comment]
These bugs are horrible to debug, especially since I can't reproduce myself. Just upgraded to 13.4, removed the integration and re-added: worked instantly. I also tried changing it to accept a specific passcode and that worked as well. Used "devices on same network"-setting and allowed nearby devices.

@ff4500 Did you get any additional stack traces as well?

**Comment by ff4500** [comment]
> Did you get any additional stack traces as well?

I didn’t check the debug logs on this particular ATV, but I will when I run it again later. Since relaxing the setting seemed to work on the one, I need to try that on the 4K ATV first.

**Comment by postlund** [comment]
I'm not entirely sure yet, but something might have changed somewhere. I can't play things with `play_url` anymore, just get error code 500. Might be connected somehow.

**Comment by postlund** [comment]
I have not found any problems myself so far. The reason why `play_url` didn't work was because I specified a password. It did work to pair, but any other action is a no-go. So I assume something must be changed for supporting this.

**Comment by ff4500** [comment]
@postlund I forgot to come back and update on the other ATV (4k). I had zero issues with setting up the integration on that ATV, although it did reject the first pin code, which _was_ entered correctly. It's working fine now, after inputting the second code displayed though.

**Comment by postlund** [comment]
@loopway @ff4500 Do you remember if any of the PINs had leading zeros in them?

**Comment by ff4500** [comment]
@postlund I did have that issue the first time I commented (on my original ATV 1080) with the leading zero on the PIN. But subsequent PINs didn't have it and it still didn't work. The most recent time (when the PIN actually worked), neither PIN had a leading zero. Even the initial one that was rejected on my first ATV.

**Comment by loopway** [comment]
@postlund After reading @ff4500 messages, I decided to give it another try yesterday. And to my surprise it actually worked! No software levels changed since the tries before, so this confuses me a little bit. The only thing I can think of which was different is that I actually took my laptop from my office to the living room where the apple tv is to enter the pin code. Could it be a timing issue? Anyway, feel free to close the issue for now, since it's working for me now. PS: The working pin had a leading zero and I entered it without it.

**Comment by postlund** [comment]
@ff4500 @loopway Ok, seems good that it worked in the end. But I have a feeling there's a hidden bug in there somewhere, not sure if it's in `pyatv` or if the Apple TV is acting up (it does that some times). I added some additional debug logs last week that maybe can help troubleshoot (I will need to release 0.6.0 and bump to that version in the component before we can see them). Lets keep the issue open a while longer to see if someone else reports in.

@ff4500 Regarding your laptop, I don't think that is a problem. Communication is via HTTP so it's not time-bound in any way (assuming you input in PIN in a "decent" time frame at least).

**Comment by postlund** [comment]
I think we can close this for now and open a new issue in case this is seen again.
